be/tx.c

```diff
@@ -740,6 +740,37 @@ M0_INTERNAL void m0_be_tx_gc_enable(struct m0_be_tx *tx,
 	tx->t_gc_param   = param;
 }
 
+M0_INTERNAL void m0_be_tx_lsn_set(struct m0_be_tx *tx,
+                                  m0_bindex_t      lsn,
+                                  m0_bindex_t      lsn_discarded)
+{
+	/*
+	 * This function is called from m0_be_tx_group_close().
+	 * When the BE tx group this transaction belongs to is being closed
+	 * the tx sm group is not locked - this is why BE_TX_LOCKED_AT_STATE
+	 * could not be used here.
+	 */
+	M0_PRE(m0_be_tx_state(tx) == M0_BTS_CLOSED);
+
+	tx->t_lsn = lsn;
+	tx->t_lsn_discarded = lsn_discarded;
+	M0_LEAVE("tx=%p t_lsn=%"PRIu64" t_lsn_discarded=%"PRIu64,
+	         tx, tx->t_lsn, tx->t_lsn_discarded);
+}
+
+M0_INTERNAL void m0_be_tx_lsn_get(struct m0_be_tx *tx,
+                                  m0_bindex_t     *lsn,
+                                  m0_bindex_t     *lsn_discarded)
+{
+	M0_ENTRY("tx=%p t_lsn=%"PRIu64" t_lsn_discarded=%"PRIu64,
+	         tx, tx->t_lsn, tx->t_lsn_discarded);
+	M0_PRE(BE_TX_LOCKED_AT_STATE(tx, (M0_BTS_LOGGED, M0_BTS_PLACED,
+	                                  M0_BTS_DONE)));
+
+	*lsn = tx->t_lsn;
+	*lsn_discarded = tx->t_lsn_discarded;
```

Comment:
Matching M0_LEAVE() is needed.

Reply:
Please clarify how `M0_LEAVE()` helps with debugging here.

Reply:
In any case I've removed it from `fdmi_int`.

Reply:
M0_ENTRY() and M0_LEAVE()/M0_RETURN() must be matched.
The M0_ENTER() prints a ">", indicating entering a function call.
The M0_LEAVE()/M0_RETURN() prints a "<", indicating leaving a function call.

They are useful when reading trace logs.

Reply:
The matching M0_ENTRY()/M0_LEAVE() is useless if nothing of interest happens between them. It just clutters the traces. You could find tons of unmatched M0_ENTRY()/M0_LEAVE() all over the code.

If you see one of them and the 2 assignments (without any lock) above or below like in this case then adding second one wouldn't add any value when reading the log.